sw/device/tests/usbdev_aon_wake_reset.c

```diff
@@ -0,0 +1,146 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/pinmux_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/lib/testing/usb_testutils.h"
+#include "sw/device/lib/testing/usb_testutils_controlep.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+/**
+ * Configuration values for USB.
+ */
+static const uint8_t config_descriptors[] = {
+    USB_CFG_DSCR_HEAD(USB_CFG_DSCR_LEN + USB_INTERFACE_DSCR_LEN, 1),
+    // Single interface with no endpoint, just to make the host happy.
+    VEND_INTERFACE_DSCR(0, 0, 0x50, 0),
+};
+
+/**
+ * USB device context types.
+ */
+static usb_testutils_ctx_t usbdev;
+static usb_testutils_controlep_ctx_t usbdev_control;
+
+/**
+ * Pinmux handle
+ */
+static dif_pinmux_t pinmux;
+
+OTTF_DEFINE_TEST_CONFIG();
+
+static void init_peripherals(void) {
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+  pinmux_testutils_init(&pinmux);
+  CHECK_DIF_OK(dif_pinmux_input_select(
+      &pinmux, kTopEarlgreyPinmuxPeripheralInUsbdevSense,
+      kTopEarlgreyPinmuxInselIoc7));
+
+  CHECK_STATUS_OK(usb_testutils_init(&usbdev, /*pinflip=*/false,
+                                     /*en_diff_rcvr=*/true,
+                                     /*tx_use_d_se0=*/false));
+  CHECK_STATUS_OK(usb_testutils_controlep_init(
+      &usbdev_control, &usbdev, 0, config_descriptors,
+      sizeof(config_descriptors), NULL, 0));
+}
+
+// Enable/disable the AON Wake module, and wait with timeout until that has
+// been confirmed.
+static status_t aon_wait(dif_toggle_t enable) {
+  // We must be sure that any alteration to the USBDEV pull up enables has
+  // propagated through the CDC synchronizer and been sampled on the lower
+  // frequency (200kHz) AON clock; allow 3 clock cycles.
+  busy_spin_micros(15u);
+  TRY(dif_usbdev_set_wake_enable(usbdev.dev, enable));
+  // The AON Wake module operates on a 200kHz clock, so the clock period is
+  // 5us; we have CDC between USBDEV and AON Wake, but it responds within a
+  // couple of its clock cycles, so this is plenty.
+  ibex_timeout_t timeout = ibex_timeout_init(20);
+  do {
+    dif_usbdev_wake_status_t status;
+    TRY(dif_usbdev_get_wake_status(usbdev.dev, &status));
+    // In the requested state yet?
+    if (status.active == dif_toggle_to_bool(enable)) {
+      return OK_STATUS();
+    }
+  } while (!ibex_timeout_check(&timeout));
+
+  return DEADLINE_EXCEEDED();
+}
+
+static void wait_until_configured(uint32_t timeout_micros) {
```

Comment:
Suggestion: return 'status_t' and check in the parent, as per these other utility functions, please.
This function will be useful in other tests, and will probably become a shared cc_library at some point.